### ISSUE TITLE
Use router.replace to forget _load

### ISSUE THREAD

**Issue by Kruptein** [Kruptein/PlanarAlly#199]
I currently use a _load temporary route to show while prefetching stuff.  This ends up in the window history.

This can be remedied by using router replace.

**Comment by ZachMyers3** [comment]
What I've done in Vue projects is keep a loading flag and nest my component's ```<div>``` within a ```v-if```. I have a simple example below.

You could potentially have a loading state in the store, and then do something similar?

Just throwing this out there. This has allowed me to lazy load some components fairly easily with seamless loading built in.

```vue
<template>
    <div v-if="!loading">
        <!-- component template here -->
    </div>
    <div v-else>
        <!-- circular loading overlay -->
        <v-overlay>
            <v-progress-circular
                indeterminate
                size="64"
            ></v-progress-circular>
        </v-overlay>
    </div>
</template>
```

**Comment by Kruptein** [comment]
This might indeed be a more elegant approach

**Comment by ZachMyers3** [comment]
Trying this locally, trying to get loading to be based for ```coreStore.initialized```. I can get routing and authentication to work properly with my router adjustments but the game map isn't loading. Curious if you might know more. New router is below.

```typescript
// route.ts
router.beforeEach(async (to, _from, next) => {
    if (to.matched.some(record => record.meta.auth)) {
        // authentication is required, but the store isnt authenticated
        if (!coreStore.initialized) {
            console.log("Not authenticated");
            // verify with the api to see if really authenticated
            const response = await fetch("/api/auth");
            if (response.ok) {
                const data = await response.json();
                // api reports authentication, adjust state and proceed
                coreStore.setAuthenticated(true);
                coreStore.setUsername(data.username);
                coreStore.setEmail(data.email);
                coreStore.setInitialized(true);
                next();
            } else {
                // truly not authenticated, redirect to login
                next({
                    path: "/auth/login",
                    query: {
                        redirect: to.path,
                    },
                });
            }
        } else {
            // the store says client is initialized
            next();
        }
    } else {
        // no authentication required
        next();
    }
});
```

**Comment by Kruptein** [comment]
Just at a glance there are some checks that are dropped from the original code.
You no longer check the `coreStore.authenticated` field for the auth records. As soon as the coreStore is initialized the if check will be skipped entirely and no auth is verified.
The reason this currently works is because the data.auth check is also skipped upon receiving data from the `/api/auth` call.  The returned response can be 200 but with a failed authentication.

I don't think this in particular would mess up the game board loading, but should should be addressed nonetheless.

**Comment by ZachMyers3** [comment]
I was originally checking coreStore.authenticated but tried to match the
old logic a little closer.

Are you saying the API can return an OK response but the client not be
authenticated?

**Comment by Kruptein** [comment]
Yes.  The server logic is

```python
async def is_authed(request):
    user = await authorized_userid(request)
    if user is None:
        data = {"auth": False, "username": ""}
    else:
        data = {"auth": True, "username": user.name, "email": user.email}
    return web.json_response(data)

**Comment by ZachMyers3** [comment]
I was just assuming it was sending back a 401, that's my fault.

I'll refactor it to account for a false authentication as well. I was thinking of only handling authentication inside of the router and then handling the initialized state in my loading logic inside app.vue. I'll play with it a little bit and report back.

**Comment by ZachMyers3** [comment]
Here's what I've got on my ```router``` branch on my fork.

```typescript
// router.ts
router.beforeEach(async (to, _from, next) => {
    if (to.matched.some(record => record.meta.auth)) {
        // authentication is required, but the store isnt authenticated
        if (!coreStore.authenticated) {
            console.log("Not authenticated");
            // verify with the api to see if really authenticated
            const response = await fetch("/api/auth");
            if (response.ok) {
                const data = await response.json();
                if (data.auth) {
                    // api reports authentication, adjust state and proceed
                    coreStore.setAuthenticated(true);
                    coreStore.setUsername(data.username);
                    coreStore.setEmail(data.email);
                    coreStore.setInitialized(true);
                    next();
                } else {
                    // api reports not authenticated, redirect
                    next({
                        path: "/auth/login",
                        query: {
                            redirect: to.path,
                        },
                    });
                }
            } else {
                // truly not authenticated, redirect to login
                next({
                    path: "/auth/login",
                    query: {
                        redirect: to.path,
                    },
                });
            }
        } else {
            // the store says client is initialized
            next();
        }
    } else {
        // no authentication required
        next();
    }
});
```

I've got routing working as it did now, there's just no loading message. I think I'll remove ```setInitialized()``` and place that inside of app.vue.

**Comment by ZachMyers3** [comment]
Got it figured out! 

I was able to remove the _load directory from the router and move the concept into a loading state in the core store. Now on ```router.beforeEach()``` this loading state gets enabled and then ```router.afterEach()``` disables the state. From there I was able to watch the state from within ```App.vue``` and use <v-if> and <v-else> to display LoadComponent given the state.

Below are the key changes, if this looks good to you I can create a PR for it. Currently it works as a drop-in replacement as the current loading scheme.

```vue
// App.vue
<template>
    <div id="app">
        <LoadComponent v-if="loading"></LoadComponent>
        <router-view v-else ref="activeComponent"></router-view>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import LoadComponent from "@/core/components/load.vue";
import { coreStore } from "@/core/store";

@Component({
    components: {
        LoadComponent,
    },
})
export default class App extends Vue {
    get loading(): boolean {
        return coreStore.loading;
    }
}
</script>
```

```typescript
// router.ts
router.beforeEach(async (to, _from, next) => {
    coreStore.setLoading(true);
    if (to.matched.some(record => record.meta.auth)) {
        // authentication is required, but the store isnt authenticated
        if (!coreStore.authenticated) {
            console.log("Not authenticated");
            // verify with the api to see if really authenticated
            const response = await fetch("/api/auth");
            if (response.ok) {
                const data = await response.json();
                if (data.auth) {
                    // api reports authentication, adjust state and proceed
                    coreStore.setAuthenticated(true);
                    coreStore.setUsername(data.username);
                    coreStore.setEmail(data.email);
                    coreStore.setInitialized(true);
                    next();
                } else {
                    // api reports not authenticated, redirect
                    next({
                        path: "/auth/login",
                        query: {
                            redirect: to.path,
                        },
                    });
                }
            } else {
                // truly not authenticated, redirect to login
                next({
                    path: "/auth/login",
                    query: {
                        redirect: to.path,
                    },
                });
            }
        } else {
            // the store says client is initialized
            next();
        }
    } else {
        // no authentication required
        next();
    }
});

router.afterEach((_to, _from) => {
    coreStore.setLoading(false);
});
```

```typescript
// store.ts
class CoreStore extends VuexModule implements CoreState {
    // ...
    loading = false;
    // ...
    @Mutation
    setLoading(loading: boolean): void {
        this.loading = loading;
    }
```

**Comment by Kruptein** [comment]
You're always free to start a pr btw, just make sure to make it clear in the description that it's a work in progress.  I'm not sure if I'll have time to check it out tonight, but I'll try to respond somewhere this weekend.

**Comment by Kruptein** [comment]
Closed in pr #215
